fdbserver/workloads/WriteTagThrottling.actor.cpp

```diff
@@ -0,0 +1,228 @@
+ * limitations under the License.
+ */
+
+#include "fdbserver/TesterInterface.actor.h"
+#include "fdbserver/workloads/workloads.actor.h"
+#include "fdbserver/WorkerInterface.actor.h"
+#include "fdbclient/NativeAPI.actor.h"
+#include "fdbclient/TagThrottle.h"
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+// workload description:
+// This workload aims to test whether we can throttling some bad clients that doing penetrating write on write hot-spot
+// range. There are several good clientActor just randomly do read and write ops in transaction. Also, some bad
+// clientActor has high probability to read and write a particular hot-spot. If the tag-based throttling works right on
+// write-heavy tags, it will only limit the bad clientActor without influence other normal actors. This workload also
+// output TPS and latency of read/set/clear operations to do eyeball check. We want this new feature would not cause
+// more load to the cluster (Maybe some visualization tools is needed to show the trend of metrics)
+struct WriteTagThrottlingWorkload : TestWorkload {
+	// Performance metrics
+	PerfIntCounter transactions, retries, tooOldRetries, commitFailedRetries;
+	PerfDoubleCounter totalLatency, readOpLatency, commitOpLatency;
+	double clientReadLatency = 0, worstReadLatency = 0;
+	double clientCommitLatency = 0, worstCommitLatency = 0;
+
+	// Test configuration
+	int actorPerClient, goodActorPerClient, badActorPerClient;
+	int numWritePerTx, numReadPerTx, numClearPerTx;
+	int keyCount;
+	double badOpRate;
+	double testDuration;
+	bool writeThrottle;
+
+	// internal states
+	Key readKey, clearKey;
+	int clearKeyInt;
+	double txBackoffDelay;
+	TransactionTag myTag;
+	static constexpr const char* NAME = "WriteTagThrottling";
+
+	WriteTagThrottlingWorkload(WorkloadContext const& wcx)
+	  : TestWorkload(wcx), transactions("Transactions"), retries("Retries"), totalLatency("Latency"),
+	    tooOldRetries("Retries.too_old"), commitFailedRetries("Retires.commit_failed"), readOpLatency("ReadOpLatency"),
+	    commitOpLatency("CommitOpLatency") {
+		testDuration = getOption(options, LiteralStringRef("testDuration"), 120.0);
+		badOpRate = getOption(options, LiteralStringRef("badOpRate"), 0.9);
+		numWritePerTx = getOption(options, LiteralStringRef("numWritePerTx"), 1);
+		numReadPerTx = getOption(options, LiteralStringRef("numReadPerTx"), 1);
+		numClearPerTx = getOption(options, LiteralStringRef("numClearPerTx"), 1);
+
+		writeThrottle = getOption(options, LiteralStringRef("writeThrottle"), false);
+		actorPerClient = getOption(options, LiteralStringRef("actorPerClient"), 10);
+		badActorPerClient = getOption(options, LiteralStringRef("badActorPerClient"), 1);
+		goodActorPerClient = actorPerClient - badActorPerClient;
+		keyCount = getOption(options, LiteralStringRef("keyCount"), 100);
+		readKey = StringRef(format("testKey%010d", deterministicRandom()->randomInt(0, keyCount / 2)));
```

Comment:
You can inherit from the `KVWorkload` class and it will provide helper functions with similar functionality.